src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
 {
-    colMapper_ = -1;
+    swap(lhs.gt_, rhs.gt_);
+
+    std::swap(lhs.max_evects_, rhs.max_evects_);
+    std::swap(lhs.spect_tol_, rhs.spect_tol_);
+
+    swap(lhs.P_edge_, rhs.P_edge_);
+    swap(lhs.P_vertex_, rhs.P_vertex_);
+    swap(lhs.face_cdof_, rhs.face_cdof_);
+    swap(lhs.agg_bubble_dof_, rhs.agg_bubble_dof_);
+
+    swap(lhs.vertex_targets_, rhs.vertex_targets_);
+    swap(lhs.edge_targets_, rhs.edge_targets_);
+    swap(lhs.agg_ext_sigma_, rhs.agg_ext_sigma_);
+
+    swap(lhs.col_marker_, rhs.col_marker_);
+
+    swap(lhs.B_potential_, rhs.B_potential_);
+    swap(lhs.D_trace_sum_, rhs.D_trace_sum_);
+    swap(lhs.D_trace_, rhs.D_trace_);
+    swap(lhs.F_potential_, rhs.F_potential_);
 }
 
-void GraphCoarsen::BuildPVertices(
-    std::vector<mfem::DenseMatrix>& vertex_target,
-    mfem::SparseMatrix& Pvertices, bool build_coarse_relation)
+void GraphCoarsen::ComputeVertexTargets(const ParMatrix& M_ext_global,
+                                        const ParMatrix& D_ext_global)
 {
-    // simply put vertex_target into Pvertices
+    const SparseMatrix& M_ext = M_ext_global.GetDiag();
+    const SparseMatrix& D_ext = D_ext_global.GetDiag();
+
+    int num_aggs = gt_.agg_ext_edge_.Rows();
 
-    const unsigned int nAggs = vertex_target.size();
-    const mfem::SparseMatrix& Agg_vertex(graph_topology_.Agg_vertex_);
-    int nvertices = Agg_vertex.Width();
-    int nlocal_fine_dofs, nlocal_coarse_dofs;
-    mfem::Array<int> local_fine_dofs;
+    DenseMatrix evects;
+    LocalEigenSolver eigs(max_evects_, spect_tol_);
 
-    int* Pvertices_i = new int[nvertices + 1];
-    Pvertices_i[0] = 0;
-    for (unsigned int i = 0; i < nAggs; ++i)
+    for (int agg = 0; agg < num_aggs; ++agg)
     {
-        GetTableRow(Agg_vertex, i, local_fine_dofs);
-        nlocal_coarse_dofs = vertex_target[i].Width();
-        nlocal_fine_dofs = local_fine_dofs.Size();
-        for (int j = 0; j < nlocal_fine_dofs; ++j)
-            Pvertices_i[local_fine_dofs[j] + 1] = nlocal_coarse_dofs;
+        std::vector<int> edge_dofs_ext = GetExtDofs(gt_.agg_ext_edge_, agg);
+        std::vector<int> vertex_dofs_ext = GetExtDofs(gt_.agg_ext_vertex_, agg);
+        std::vector<int> vertex_dofs_local = gt_.agg_vertex_local_.GetIndices(agg);
+
+        if (edge_dofs_ext.size() == 0)
+        {
+            vertex_targets_[agg] = DenseMatrix(1, 1, {1.0});
+            continue;
+        }
+
+        SparseMatrix M_sub = M_ext.GetSubMatrix(edge_dofs_ext, edge_dofs_ext, col_marker_);
+        SparseMatrix D_sub = D_ext.GetSubMatrix(vertex_dofs_ext, edge_dofs_ext, col_marker_);
+        SparseMatrix D_sub_T = D_sub.Transpose();
+
+        D_sub_T.InverseScaleRows(M_sub);
+
+        SparseMatrix DMinvDT = D_sub.Mult(D_sub_T);
+
+        eigs.Compute(DMinvDT, evects);
+
+        if (evects.Cols() > 1)
+        {
+            DenseMatrix evects_ortho = evects.GetCol(1, evects.Cols());
+            agg_ext_sigma_[agg] = D_sub_T.Mult(evects_ortho);
```

Comment:
This can reduce to one line (and avoid copying) if you can add something like the idea of `OffsetMult` in `GraphEdgeSolver` to `DenseMatrix`, but it's up to you (only if you think this kind of function will in useful in general for the linear algebra library).

Reply:
Yes, `OffsetMult` can be added to `Operator` class since it doesn't really matter what is applying the `Mult`.